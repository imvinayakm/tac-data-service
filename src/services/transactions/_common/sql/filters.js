const { where, whereIn, limit } = require('../../../../utils/db/knex/index');

const id = where('id');
const ids = whereIn('id');

const sender = where('sender');

const senders = whereIn('sender');

const timeStart = where('time_stamp', '>=');
const timeEnd = where('time_stamp', '<=');

const sort = s => q =>
  q
    .clone()
    .orderBy('time_stamp', s)
    .orderBy('id', s);

const after = ({ timestamp, id, sort }) => q => {
  const comparator = sort === 'desc' ? '<' : '>';
  return q
    .clone()
    .whereRaw(`(time_stamp, id) ${comparator} (?, ?)`, [timestamp, id]);
};

module.exports = {
  id,
  ids,
  sender,
  senders,
  timeStart,
  timeEnd,
  sort,
  after,
  limit,
};
