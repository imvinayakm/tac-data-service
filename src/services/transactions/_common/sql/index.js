const {
  compose,
  map,
  pick,
  filter,
  has,
  __,
  identity,
  reverse,
  merge,
} = require('ramda');

const commonFiltersOrder = require('./filtersOrder');
const commonFilters = require('./filters');

const createSql = ({
  query,
  filters = commonFilters,
  filtersOrder = commonFiltersOrder,
  queryAfterFilters = {},
}) => {
  const queryAfterFiltersWithDefaults = merge(
    {
      get: identity,
      mget: identity,
      search: identity,
    },
    queryAfterFilters
  );

  return {
    get: id =>
      compose(
        String,
        q => queryAfterFiltersWithDefaults.get(q, id),
        filters.id(id)
      )(query),

    mget: ids =>
      compose(
        String,
        q => queryAfterFiltersWithDefaults.mget(q, ids),
        filters.ids(ids)
      )(query),

    search: fValues => {
      const fValuesPicked = pick(filtersOrder, fValues);
      const appliedFs = compose(
        map(x => filters[x](fValuesPicked[x])),
        filter(has(__, fValuesPicked)),
        reverse
      )(filtersOrder);

      return compose(
        String,
        q => queryAfterFiltersWithDefaults.search(q, fValuesPicked),
        ...appliedFs,
        q => q.clone()
      )(query);
    },
  };
};

module.exports = createSql;
