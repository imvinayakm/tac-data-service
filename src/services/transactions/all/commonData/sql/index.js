const createSql = require('../../../_common/sql');

const { select } = require('./query');
const { filters, filtersOrder } = require('./filters')

module.exports = createSql({
  query: select,
  filters,
  filtersOrder
});
