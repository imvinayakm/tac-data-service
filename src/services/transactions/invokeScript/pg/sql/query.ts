import * as knex from 'knex';
const pg = knex({ client: 'pg' });

const columnsWithoutFeeAndPaymentAssetId = [
  // common
  'height',
  'tx_type',
  'id',
  'time_stamp',
  'signature',
  'proofs',
  'tx_version',
  'status',
  'sender',
  'sender_public_key',

  // type-specific
  'dapp',
  'function_name',

  // args
  'arg_type',
  'arg_value_integer',
  'arg_value_boolean',
  'arg_value_binary',
  'arg_value_string',
  'arg_value_list',
  'position_in_args',

  // payment
  // payment asset id queries at select function
  'amount',
  'position_in_payment',
];

export const txs = pg({ t: 'txs_16' })
  .select('*')
  .leftJoin({ a: 'txs_16_args' }, 'a.tx_id', 't.id')
  .leftJoin({ p: 'txs_16_payment' }, 'p.tx_id', 't.id');

export const fSelect = pg.select('id').from('txs_16');

export const select = pg
  .select(columnsWithoutFeeAndPaymentAssetId)
  .select({
    amount: pg.raw(
      'amount * 10^(case when txs.asset_id is null then -8 else -p_dec.decimals end)'
    ),
    asset_id: 'txs.asset_id',
  })
  .select({
    fee: pg.raw('fee * 10^(-8)'),
  })
  .from({ txs })
  .leftJoin({ p_dec: 'asset_decimals' }, 'p_dec.asset_id', 'txs.asset_id');

export const composeQuery = (
  filteringQ: knex.QueryBuilder
): knex.QueryBuilder =>
  select.clone().whereIn('id', pg.select('id').from({ filtered: filteringQ }));
