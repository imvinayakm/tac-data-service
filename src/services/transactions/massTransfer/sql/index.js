const { compose } = require('ramda');

const createSql = require('../../_common/sql/index');

const { select, withTransfersDecimalsAndGrouping } = require('./query');
const { filters, filtersOrder } = require('./filters');

const queryAfterFilters = {
  get: withTransfersDecimalsAndGrouping,
  mget: withTransfersDecimalsAndGrouping,
  search: (q, fValues) =>
    compose(
      filters.sort(fValues.sort),
      withTransfersDecimalsAndGrouping
    )(q),
};

module.exports = createSql({
  query: select,
  filters,
  filtersOrder,
  queryAfterFilters,
});
