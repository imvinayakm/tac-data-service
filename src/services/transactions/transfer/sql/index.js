const { compose } = require('ramda');

const createSql = require('../../_common/sql/index');

const { select, withDecimals } = require('./query');
const { filters, filtersOrder } = require('./filters');

const queryAfterFilters = {
  get: withDecimals,
  mget: withDecimals,
  search: (q, fValues) =>
    compose(
      filters.sort(fValues.sort),
      withDecimals
    )(q),
};

module.exports = createSql({
  query: select,
  filters,
  filtersOrder,
  queryAfterFilters,
});
