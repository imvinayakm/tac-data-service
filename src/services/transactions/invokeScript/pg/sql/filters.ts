import * as knex from '../../../../../utils/db/knex';

import * as commonFilters from '../../../_common/sql/filters';

export default {
  ...commonFilters,
  dapp: knex.where('dapp'),
  function: knex.where('function_name'),
};
