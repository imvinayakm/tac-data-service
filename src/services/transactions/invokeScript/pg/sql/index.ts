import { createApi } from './api';
import filters from './filters';

export default createApi({ filters });
