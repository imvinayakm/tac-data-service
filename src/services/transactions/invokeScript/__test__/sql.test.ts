import sql from '../pg/sql';
const filterValues = [
  {
    dapp: '3N92gNUHh6jnZmZtdgECTm3cNAZb1Zei6Ve',
  },
  {
    function: 'deposit',
  },
  {
    dapp: 'b3N92gNUHh6jnZmZtdgECTm3cNAZb1Zei6Ve',
    function: 'deposit',
  },
];

describe('Sql search by type-specific filters', () => {
  it('supports dapp filter', () => {
    expect(sql.search(filterValues[0])).toMatchSnapshot();
  });

  it('supports function filter', () => {
    expect(sql.search(filterValues[1])).toMatchSnapshot();
  });

  it('supports dapp and function filters', () => {
    expect(sql.search(filterValues[2])).toMatchSnapshot();
  });
});
