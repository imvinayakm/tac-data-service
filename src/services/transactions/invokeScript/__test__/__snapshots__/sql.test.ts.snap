// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sql search by type-specific filters supports dapp and function filters 1`] = `"select \\"height\\", \\"tx_type\\", \\"id\\", \\"time_stamp\\", \\"signature\\", \\"proofs\\", \\"tx_version\\", \\"sender\\", \\"sender_public_key\\", \\"dapp\\", \\"function_name\\", \\"arg_type\\", \\"arg_value_integer\\", \\"arg_value_boolean\\", \\"arg_value_binary\\", \\"arg_value_string\\", \\"position_in_args\\", \\"amount\\", \\"position_in_payment\\", amount * 10^(case when txs.asset_id is null then -8 else -p_dec.decimals end) as \\"amount\\", \\"txs\\".\\"asset_id\\" as \\"asset_id\\", fee * 10^(-8) as \\"fee\\" from (select * from \\"txs_16\\" as \\"t\\" left join \\"txs_16_args\\" as \\"a\\" on \\"a\\".\\"tx_id\\" = \\"t\\".\\"id\\" left join \\"txs_16_payment\\" as \\"p\\" on \\"p\\".\\"tx_id\\" = \\"t\\".\\"id\\") as \\"txs\\" left join \\"asset_decimals\\" as \\"p_dec\\" on \\"p_dec\\".\\"asset_id\\" = \\"txs\\".\\"asset_id\\" where \\"id\\" in (select \\"id\\" from (select \\"id\\" from \\"txs_16\\" where \\"dapp\\" = 'b3N92gNUHh6jnZmZtdgECTm3cNAZb1Zei6Ve' and \\"function_name\\" = 'deposit' order by \\"time_stamp\\" desc, \\"id\\" desc limit 100) as \\"filtered\\") order by \\"time_stamp\\" desc, \\"id\\" desc"`;

exports[`Sql search by type-specific filters supports dapp filter 1`] = `"select \\"height\\", \\"tx_type\\", \\"id\\", \\"time_stamp\\", \\"signature\\", \\"proofs\\", \\"tx_version\\", \\"sender\\", \\"sender_public_key\\", \\"dapp\\", \\"function_name\\", \\"arg_type\\", \\"arg_value_integer\\", \\"arg_value_boolean\\", \\"arg_value_binary\\", \\"arg_value_string\\", \\"position_in_args\\", \\"amount\\", \\"position_in_payment\\", amount * 10^(case when txs.asset_id is null then -8 else -p_dec.decimals end) as \\"amount\\", \\"txs\\".\\"asset_id\\" as \\"asset_id\\", fee * 10^(-8) as \\"fee\\" from (select * from \\"txs_16\\" as \\"t\\" left join \\"txs_16_args\\" as \\"a\\" on \\"a\\".\\"tx_id\\" = \\"t\\".\\"id\\" left join \\"txs_16_payment\\" as \\"p\\" on \\"p\\".\\"tx_id\\" = \\"t\\".\\"id\\") as \\"txs\\" left join \\"asset_decimals\\" as \\"p_dec\\" on \\"p_dec\\".\\"asset_id\\" = \\"txs\\".\\"asset_id\\" where \\"id\\" in (select \\"id\\" from (select \\"id\\" from \\"txs_16\\" where \\"dapp\\" = '3N92gNUHh6jnZmZtdgECTm3cNAZb1Zei6Ve' order by \\"time_stamp\\" desc, \\"id\\" desc limit 100) as \\"filtered\\") order by \\"time_stamp\\" desc, \\"id\\" desc"`;

exports[`Sql search by type-specific filters supports function filter 1`] = `"select \\"height\\", \\"tx_type\\", \\"id\\", \\"time_stamp\\", \\"signature\\", \\"proofs\\", \\"tx_version\\", \\"sender\\", \\"sender_public_key\\", \\"dapp\\", \\"function_name\\", \\"arg_type\\", \\"arg_value_integer\\", \\"arg_value_boolean\\", \\"arg_value_binary\\", \\"arg_value_string\\", \\"position_in_args\\", \\"amount\\", \\"position_in_payment\\", amount * 10^(case when txs.asset_id is null then -8 else -p_dec.decimals end) as \\"amount\\", \\"txs\\".\\"asset_id\\" as \\"asset_id\\", fee * 10^(-8) as \\"fee\\" from (select * from \\"txs_16\\" as \\"t\\" left join \\"txs_16_args\\" as \\"a\\" on \\"a\\".\\"tx_id\\" = \\"t\\".\\"id\\" left join \\"txs_16_payment\\" as \\"p\\" on \\"p\\".\\"tx_id\\" = \\"t\\".\\"id\\") as \\"txs\\" left join \\"asset_decimals\\" as \\"p_dec\\" on \\"p_dec\\".\\"asset_id\\" = \\"txs\\".\\"asset_id\\" where \\"id\\" in (select \\"id\\" from (select \\"id\\" from \\"txs_16\\" where \\"function_name\\" = 'deposit' order by \\"time_stamp\\" desc, \\"id\\" desc limit 100) as \\"filtered\\") order by \\"time_stamp\\" desc, \\"id\\" desc"`;
